src/coreclr/jit/optcse.cpp

```diff
@@ -2604,7 +2604,13 @@ class CSE_Heuristic
             // If we don't have a lot of variables to enregister or we have a floating point type
             // then we will likely need to spill an additional caller save register.
             //
-            if ((enregCount < (CNT_CALLEE_ENREG * 3 / 2)) || varTypeIsFloating(candidate->Expr()->TypeGet()))
+            if (varTypeIsFloating(candidate->Expr()))
+            {
+                // Do not make CSE for live-across-call floating point numbers
+                return false;
+            }
```

Comment:
That seems overly conservative. It seems more like the heuristics need to be adjusted since it's over-estimating the price of reloading a float constant from memory compared to having to spill it.

Reply:
Do you mean that in some cases it's still make sense to CSE live-across-call registers? e.g.
```csharp
Console.WriteLine(x * 3.0 * x * 3.0 * x * 3.0); // 3 usages
Console.WriteLine(3.0);
```
?

Reply:
Perhaps this needs to be adjusted:
https://github.com/dotnet/runtime/blob/225acfe78d1b308a7e1105278d72aecc9ffc2e70/src/coreclr/jit/optcse.cpp#L2493-L2509

For floating point variables on SysV that are live across call we never expect to be able to enregister them.

Reply:
> Do you mean that in some cases it still makes sense to CSE live-across-call registers? e.g.

Right, I would expect that for some floating point calculations it is still worth it to CSE them and spill them to stack vs having to recompute them.

Reply:
Looks like disabling canEnregister for floats + `CNT_CALLEE_SAVED_FLOAT == 0` leads to regressions like this: https://www.diffchecker.com/VpkKWnWD

Reply:
I wouldn't expect that candidate to have `LiveAcrossCall()` as true, does it?

Reply:
@jakobbotsch oops, right, I forgot to check it